src/ronin/basefee-treasury/RoninBaseFeeTreasury.sol

```diff
@@ -0,0 +1,546 @@
+    _incrementNonce();
+
+    require(!(denom == 0 || num == 0 || num >= denom), ErrInvalidThreshold(num, denom));
+
+    _threshold = Threshold(num, denom);
+
+    emit ThresholdUpdated(msg.sender, num, denom);
+  }
+
+  /**
+   * @dev Increment the global nonce.
+   *
+   * Emits a {GlobalNonceUpdated} event.
+   */
+  function _incrementNonce() internal {
+    ++_globalNonce;
+    emit GlobalNonceUpdated(msg.sender, getGlobalNonce());
+  }
+
+  /**
+   * @dev Try to remove the pending proposals.
+   */
+  function _syncPendingProposals() internal {
+    uint256 length = _pendingProposals.length();
+    bytes32[] memory pendingProposals = _pendingProposals.values();
+
+    for (uint256 i; i < length; ++i) {
+      if (getState(pendingProposals[i]) != State.Active) {
+        _pendingProposals.remove(pendingProposals[i]);
+      }
+    }
+  }
+
+  /**
+   * @dev Calculate the minimum vote power required for a proposal to pass.
+   */
+  function _calcMinVoteFor(
+    uint256 totalVotePower
+  ) internal view returns (uint96 minVoteFor) {
+    minVoteFor = ((totalVotePower * _threshold.num) / _threshold.denom).toUint96();
+    require(minVoteFor != 0, ErrZeroMinVotePower());
+  }
+
+  /**
+   * @dev Calculate the minimum vote power required for a proposal to cancel.
+   */
+  function _calcMinVoteAgainst(
+    uint256 totalVotePower
+  ) internal view returns (uint96 minVoteAgainst) {
+    uint256 diff = _threshold.denom - _threshold.num;
+    minVoteAgainst = ((totalVotePower * diff) / _threshold.denom).toUint96();
+    require(minVoteAgainst != 0, ErrZeroMinVotePower());
+  }
+
+  /**
+   * @dev Require the proposal nonce to be equal to the global nonce.
+   */
+  function _requireValidNonce(
+    uint256 nonce
+  ) internal view {
+    require(nonce == getGlobalNonce(), ErrInvalidNonce(getGlobalNonce(), nonce));
+  }
+
+  /**
+   * @dev Require the proposal state to be equal to the required state.
+   */
+  function _requireState(bytes32 hash, State requiredState) internal view {
+    State currState = getState(hash);
+    require(currState == requiredState, ErrRequiredStateNotReached(hash, requiredState, currState));
+  }
+
+  /**
+   * @dev Require the candidate to be an admin.
+   * The candidate must not be revoked.
+   */
+  function _requireCandidateAdmin(
+    address cid
+  ) internal view {
+    require(
+      IProfile(getContract(ContractType.PROFILE)).getId2Admin(cid) == msg.sender
+        && ICandidateManager(getContract(ContractType.VALIDATOR)).getCandidateInfoById(cid).revokingTimestamp == 0,
+      ErrUnauthorized(msg.sig, RoleAccess.CANDIDATE_ADMIN)
+    );
+  }
+
+  /**
+   * @dev See {TimingStorage-_computePeriod}.
+   *
+   * This duplicates the implementation in {RoninValidatorSet-_computePeriod} to reduce external calls.
+   */
+  function _computePeriod(
+    uint256 timestamp
+  ) internal pure returns (uint256) {
+    return timestamp / PERIOD_DURATION;
+  }
```

Comment:
have seen this logic serveral times, should we move it into a library for reuse?